### DISCUSSION TITLE
C# Nullable - Usage with Generics

### DISCUSSION THREAD

**Discussion by ghost** [dotnet/csharplang#3060] [General]
So in a project or code block that enables the C# `nullable` feature, if you try to declare a generic class thus:

```
public class Foo<T> {

// consider this basic statement:
private T _instance;

}
```

you enter into world of trouble if you could want the `T` to be both a `class` and a `struct` type. For example, consider the built-in type `List<T>`. How would you safely use it with full `nullable` support? 

What is the guidance here?

Tagging @MadsTorgersen for his thoughts

**Comment by huoyaoyuan** [comment]
This has been discussed recently. See https://github.com/dotnet/csharplang/blob/master/meetings/2019/LDM-2019-11-25.md

**Comment by ghost** [comment]
So (and to @MadsTorgersen ), those discussions centre around methods and method signatures where one is able to also leverage tricks like the `[NotNull]` style of attribute annotations.

My question is on the declaration of a generic `class` type. Those attributes are disallowed there.

**Comment by mikernet** [comment]
What is your specific issue? If you show an example of what you want to represent but are having trouble with then we can show you how you should do it, but right now I can't tell what your problem is.

An unconstrained `T` means it can be anything, including nullable types. If you only want `T` to be non-nullable types, then you apply a `where T : notnull` constraint to the generic parameter, i.e:

```c#
public class MyList<T> where T : notnull
{
}
```

The thing that I'm really growing to hate right now which is throwing me off when reading code is this:

```c#
public class MyClass<T1, T2> where T1 : class
{
    public void DoSomething(T1 value1, T2 value2) { }
}
```

Looking at that method signature it looks like both are non-nullable, but that's not the case...`T1 value1` is not nullable and `T2 value2` could be. It makes reasoning about the nullability of your parameters very difficult without constantly referring back to the type signature and doing the mental math of which type definition means what kind of possible nullability, which can have a surprisingly high cognitive load for more complex classes with various combinations of generic constraints. Ugh.

I think the way nullables are implemented is going to make C# a nightmare for new developers to grasp.

**Comment by mikernet** [comment]
I'll give you a sample definition of some `List<T>` members so you can get an idea of how this works. I just went through this exercise for a similar use case, and this is the way I would do it.

First, you have to understand that an unconstrained `T` could be nullable, or it might not be. For example, all of these are valid usages of a list:

```c#
var y = new List<string>(); // non-nullable ref type
var x = new List<string?>(); // nullable ref type
var z = new List<int>(); // non-nullable value type
var w = new List<int?>(); // nullable value type
```

So, when writing the guts of your list class, you have to consider that `T` **could** be any of those depending on the way the class is used. For most cases, you can just use `T` as you normally would and it will do the right thing. Let's start with an `Add` method:

```c#
public void Add(T value)
{
    // implementation...
}
```

You can safely leave this as-is. If the user defines the list as `List<string>` and does `Add(null)` it will show a warning because the T parameter is considered non-nullable. In order to do this without a warning, the user needs to define the list as a `List<string?>`.

The one "problem" with the way NRTs are implemented which unfortunately can't be worked around and that you should be aware of is that there is no way to make unconstrained reference typed `T` values as "null safe" as you can with value typed `T` values. There is no way to prevent users from doing something like this:

```c#
var x = new List<string>();
// no way for me to detect that `T` is non-nullable inside my Add method 
// so that I can throw NullArgumentException :(
x.Add(null!); 

var y = new List<int>();
y.Add(null); // absolutely no way the user can do this with value types to add a null value
```

There is just no way to know inside your class whether a reference typed `T` is nullable or not. This is somewhat unfortunate as a bug can cause `null` to propagate to places you don't expect given that you have specified non-nullable types and then pop up later far away from the bug location as it gets passed around in variables that are supposed to be non-nullable, so keep this in mind.

Indexers will work as you expect them to without much thought:

```c#
public T this[int index]{
    get; // implement get
    set; // implement set
}
```

The complications come into play when you might return null even though the user has specified a `T` generic parameter that is not null. For example, say you want to write a method such as this:

```c#
public T GetAtIndexOrDefault(int index) 
{
    if (index >= 0 && index < _length)
        return _items[index];

    // Can't return "default" here because T might be a non-nullable string for example
    return default(T); 
}
```

The way we properly annotate this method to account for this is with the `[MaybeNull]` annotation:

```c#
[return: MaybeNull]
public T GetAtIndexOrDefault(int index)
{
    if (index >= 0 && index < _length)
        return _items[index];

    // Current compiler limitation does not recognize that return is marked MaybeNull inside
    // method body so you have to use the null forgiveness operator (!) here 
    return default(T)!; 
}
```

Now the compiler will show a warning if you try to do this:

```c#
var list = new List<string>();
string s = list.GetAtIndexOrDefault(0); // warning, return value may be null
string? s = list.GetAtIndexOrDefault(0); // no warning
```

You can get fancy with some of the other annotations to get proper dictionary-style "try" methods:

```c#
public bool TryGetAtIndex(int index, [MaybeNullWhen(false)] out T value)
{
    if (index >= 0 && index < _length) {
        value = _items[index];
        return true;
    }

    // Again, current compiler limitation does not honor [MaybeNullWhen] inside the method 
    // so ! is required
    value = default(T)!; 
    return false;
}
```

`MaybeNullWhen(false)` indicates that the output value might be `null` if the return value of the method is `false`.

`AllowNull` works in a similar way, just the other way around - it lets the user pass in a `null` value when otherwise they wouldn't be allowed to. Let's say you want to add a method `AddIfNotNull` that can safely accept a null value, even if `T` is set to a non-nullable value:

```c#
public void AddIfNotNull(T value)
{
    if (value != null)
        _items[_length++] = value;
}

var list = new List<string>();
string? value = null;
list.AddIfNotNull(value); // warning: passing a nullable value to a non-nullable parameter
```

The way we fix this is by annotating the method with `[AllowNull]` like so:

```c#
public void AddIfNotNull([AllowNull] T value)
{
    if (value != null)
        _items[_length++] = value;
}

var list = new List<string>();
string? value = null;
list.AddIfNotNull(value); // no warning
```

Hopefully that sets you off on the right path. Let me know if you have any follow-up questions.

**Comment by ghost** [comment]
@mikernet - your scenarios after the indexers in your reply is exactly what I was looking for. I was implementing a `List<T>`-style collection object and I was going to have methods in it that could return `default`/`NULL` [if the item in question was not found]. 

Let me try out your suggestions in my scenario and see how well it works or doesn't and I'll post back in a couple of days.

Thanks! and Happy New Year!!!

**Comment by bbarry** [comment]
I recently ran into the following:

    public class Foo<T>
    {
        public async Task<T> MayReturnNullIfReferenceType(...)
    }

**Comment by lostmsu** [comment]
OK, so how do I annotate a method like `FirstOrDefault<T>(...)` to get rid of `CS8653`?

**Comment by mikernet** [comment]
@lostmsu Example please, I don't know what specific situation you are running into.

@bbarry Yeah, that's unfortunately beyond the current expressiveness of NRTs. I *think* the proposal for `T??` might solve this as follows (but I'm not 100% sure, haven't fully digested that proposal):

```c#
public class Foo<T>
{
    public async Task<T??> MayReturnNullIfReferenceType(...)
}
```

**Comment by lostmsu** [comment]
```csharp
static T FirstOrDefault<T>(this IEnumerable<T> enumerable) {
  using var enumerator = enumerable.GetEnumerator();
  return enumerator.MoveNext()
    ? enumerator.Current
    // CS8653
    : default;
}

string x = FirstOrDefault(new string[0]); // should give a warning for reference types
```

Same for `value` in `IDictionary<TKey, TValue>.TryGetValue(TKey key, out TValue value)`

**Comment by Joe4evr** [comment]
> OK, so how do I annotate a method like `FirstOrDefault<T>(...)` to get rid of `CS8653`?

[You have to wait for the language team to implement `T??`.](https://github.com/dotnet/csharplang/blob/master/meetings/2019/LDM-2019-11-25.md#solutions)

**Comment by devel0** [comment]
I have problem in definition of an extension method that works either with struct and classes like the follow ( the purpose of this method is to return enumerated elements with info about the previous one, that will be null for the first element ):

```csharp
/// <summary>
/// enumerate given items returning a tuple with ( null for first hit ) prev element
/// </summary>
public static IEnumerable<(T? prev, T item)> WithPrev<T>(this IEnumerable<T> en)
{
    var enm = en.GetEnumerator();

    T prev = null;
    var item = default(T);
    while (enm.MoveNext())
    {
        item = enm.Current;
        yield return (prev, item);
        prev = item;
    }
}
```

getting compiler error

```error
A nullable type parameter must be known to be a value type or non-nullable reference type. Consider adding a 'class', 'struct', or type constraint.
```

I ended using two almost identical method implementation but I wondering if there is something that I missed to make it work with only a single method ( I already set `<LangVersion>8</LangVersion>` in csproj )

```csharp
/// <summary>
/// enumerate given items returning a tuple with ( nullable for first hit ) prev element
/// </summary>
public static IEnumerable<(T? prev, T item)> WithPrevPrimitive<T>(this IEnumerable<T> en) where T : struct
{
    var enm = en.GetEnumerator();

    T? prev = null;
    var item = default(T);
    while (enm.MoveNext())
    {
        item = enm.Current;
        yield return (prev, item);
        prev = item;
    }
}

/// <summary>
/// enumerate given items returning a tuple with ( null for first hit ) prev element
/// </summary>
public static IEnumerable<(T prev, T item)> WithPrev<T>(this IEnumerable<T> en) where T : class
{
    var enm = en.GetEnumerator();

    T prev = null;
    var item = default(T);
    while (enm.MoveNext())
    {
        item = enm.Current;
        yield return (prev, item);
        prev = item;
    }
}
```

**Comment by mikernet** [comment]
@devel0 Unfortunately no, you're running into the same issue as above that requires `T??` to be implemented. With that you can do:
```c#
public static IEnumerable<(T?? prev, T item)> ....
```

**Comment by jnm2** [comment]
It looks like `T??` is not the current plan.

https://github.com/dotnet/csharplang/blob/master/meetings/2020/LDM-2020-06-17.md#t
https://github.com/dotnet/csharplang/blob/master/meetings/2020/LDM-2020-07-01.md#Non-defaultable-struct-types-and-parameterless-struct-constructors

**Comment by Joe4evr** [comment]
Which is very sad. We sorely need a solution for type parameters in this space.

**Comment by devel0** [comment]
from [conclusion](https://github.com/dotnet/csharplang/blob/master/meetings/2020/LDM-2020-06-17.md#t)

```csharp
override void M1<T,U>(T? x) where T: class|struct    // g
```

would a resonable solution for my case in order to avoid twice code writing of the same function and would not require introduction of T?? but only the "OR" type ( class | struct )

**Comment by 333fred** [comment]
Well, we're planning on using `where T : default`. To be clear, `T??` as a syntax is dead because the syntax is truly ambiguous. Annotating type parameters the feature is not dead.

**Comment by gulshan** [comment]
I still think the title "Nullable reference types" does not convey the actual meaning of the feature. The nullability was introduced as an attribute of the references and has nothing to do with types. On the other hand, for nullable structs, it was actually part of the type system. In my opinion the better name would be- "Nullable references". This would still permit `?` notation with references within declarations of variables, fields, properties and methods. But it would disprove using `?` in type parameters. But that ship has sailed already.

**Comment by HaloFour** [comment]
@gulshan 

I agree, the name implies that it affects the type system and I think that's a common source of confusion (and angst) around the feature.  I'll use the NRT acronym but I think "nullable references" better conveys what the feature is attempting to accomplish and how it has been implemented.

**Comment by sjb-sjb** [comment]
Speaking as a user of the language (rather than a language expert or compiler expert), constructs such as [return: MaybeNull] T GetAtIndexOrDefault(--) seem quite illogical. On the one hand if T is (annotated as) a non-nullable type, then we are effectively overriding the caller's annotation of the type and jamming a null into a place where it should not be. On the other hand if T is nullable then the MaybeNull annotation is redundant. 

The inability to refer to T? where T: notnull seems like an implementation gap for the language. The meaning of T? here is perfectly sound. The fact that the implementations differ if T is a struct vs a class is something that as a language user I feel the compiler should be able to handle (with a sufficient amount of work :-). 

The two points are linked because if we could refer to T? then we could return T? from GetAtIndexOrDefault after constraining T: notnull. If we want a language that can do this without constraining T: notnull then we are saying we want a language that treats T?? the same as T? -- which is in effect what [MaybeNull] T does when T is already nullable. 

As a broader comment...I'm sure it has been discussed but I wonder why we didn't take a more type-based approach to nullability -- such as making non-nullable references into a full-fledged type and then extending Nullable<T> to include T being a non-nullable reference type (of course implementation may vary in the two cases). I take the point about backward compatibility but couldn't that be dealt with by annotating existing code, and having new code use the full type system? Fast forward 5 years and we would have a cleaner language without a lot of messy annotations everywhere.

**Comment by mikernet** [comment]
@sjb-sjb In terms of addressing why a type-based approach was not taken - the approach had to be compatible with existing API method signatures. If `Nullable<T>` was used for ref types then the entire BCL API would need to be redone with different types which would be a massive breaking change. There's also no way to "gradually" opt into NRT usage with that approach.

I agree that some aspects of it are a bit clunky but a lot of the design decisions were constrained by backwards compat requirements.

**Reply by sjb-sjb** [reply]
Not suggesting that Nullable&lt;T&gt; be used for ref types. See my [other comments](https://github.com/dotnet/csharplang/issues/3297).

**Comment by sjb-sjb** [comment]
Well, I doubt that I'm adding value by rehashing decisions that have already been made, particularly since I'm not a compiler or language expert at all. But it's so hard to resist! 

Creating a new API on top of the old API using 'real' nullable types might be a massive effort but it would not be a massive breaking change, because (a) existing code could still be compiled against the old API libraries using a compiler switch, and (b) if necessary, new code could call into the old API by using T? everywhere instead of T. True, the latter approach would not reap the benefits of nullability but then again that's what the new API layer would be for.   For example the new API would have T? TryGet() instead of bool TryGet( out T result), implementation would be the obvious one. The old API would still be there, just mapped to bool TryGet( out T? result) in the new language. A new namespace would take care of any naming conflicts with the old API. 

The payback would be that eventually the old API would deprecated and we would be in a world where new code was based on a real type system instead of having attributes all over the place that override the caller's intended typing willy-nilly. T? would actually be a nullable type and T would actually be a non-nullable type instead of just being something that looks like a non-nullable type but which may actually contain nulls. 

Again, highly hypothetical and counterfactual. Nice to dream about though.

**Comment by CyrusNajmabadi** [comment]
> True, the latter approach would not reap the benefits of nullability

That's why we didn't go that path.  The point is to get the benefit for existing code.

**Comment by sjb-sjb** [comment]
Hopefully all existing code already works :-) and the new interface could be adopted gradually during maintenance. 

Hey, it's not too late, we could still make the transition in, say, C# version 12 :-)

**Comment by CyrusNajmabadi** [comment]
the point of nullable is not about new APIs.  If you have a new API you can just design a system for it that avoids null entirely (i.e. having an Option type that gives you teh result).  The point of nullable is to annotate teh *existing* apis people use and the existing code they have so you can get reasonable warnings about the code you likely need to change because it isn't null-safe and has a decent chance of potentially crashing :)

**Comment by sjb-sjb** [comment]
Well, after a few days of hacking around with it, I've run into the same problem twice. I have an object? and an unconstrained type T. I want to cast the object? to T and have it throw an exception if the cast is not appropriate. Specifically I want to throw an exception if T is instantiated to a reference type that is attributed as non-nullable and the object? is null. But I can't figure out how to thrown an exception in that situation. The nullability attribute does not seem to be accessible based on just T. 

An example where this happens is in implementing ICommand.Execute( object? parameter) in a subclass where the parameter  type is given by a generic type. It's true that one could constrain the type to be either class or struct, but if you do that then you either have to (a) accept a big restriction on the types of parameters or else (b) you have to have duplicate copies (and I mean really textually identical) of the generic class, one for struct parameters and one for class parameters but with different names, and choose the correct one in your source code depending on the parameter type. Neither option seems appealing at all.

How should one approach this type of problem? 

P.S. Another way to state this challenge is, more broadly, how can I enforce non-nullability? How can I make sure that there is never a null in a reference-type variable that the compiler knows is non-nullable?

**Comment by mikernet** [comment]
@sjb-sjb You can't unfortunately. I've run into this many times as well and it's a limitation of the way NRTs are implemented that will have to be accepted. The method has no way of knowing whether `T` is nullable ref type or a non-nullable ref type and that's something you will have to design around.

There are typically acceptable (albeit admittedly non-optimal) ways to alter your design such that it mitigates this issue.

**Comment by mikernet** [comment]
Regarding "enforcing" non-nullability - other than extensive unit testing, you can use my [Runtime Nullables](https://github.com/Singulink/RuntimeNullables) weaver to automatically inject nullability contract tests into method/property entry and exit points for both inputs and outputs. It has definitely saved my butt a few times now, making what would otherwise be very hard to trace bugs instantly apparent :) It's not a perfect solution, i.e. it can't check field or local variable assignments, but it's pretty close.

**Comment by sjb-sjb** [comment]
That's a rockin' add-in, for sure! I just wish that the language enforced the typing instead.

**Comment by eduherminio** [comment]
I just ran into a related issue and I'm unsure about how to proceed given the current state of the language, I'd be very grateful if you could point me in the right direction:

I'm modelling a generic `Node` class with an `Id` and an 'optional'  `ParentId`. Something like:
```csharp
public class Node<TKey>     // TKey must support string, int, DateTime, etc.
{
	public TKey Id { get; set; }
	
	public TKey ParentId { get; set; }

	public Node(TKey id)
	{
		if (id?.Equals(default) == true)
		{
			throw new ArgumentException("Id cannot be null or default");
		}

		Id = id;
	}

	public bool SampleMethod(TKey key) => Id.Equals(key);  // CS8602

        // ...
}
```

I managed to avoid all warnings by constraining `TKey` to `notnull` and initializing `ParentId` to `default!`, but that doesn't seem right to me is and if I'm very probably missing a more recommended option.

**Reply by 333fred** [reply]
In C# 9, you can make ParentId TKey?, which will have the same behavior but without the warnings.

**Reply by daiplusplus** [reply]
Until you can use C# 9.0, there's two ugly workarounds:

1. Add a `bool hasParent` field and change `public TKey ParentId { get; set; }` to `public bool TryGetParent( out TKey parentId ) => parentId = this.hasParent ? default : this.ParentId`.
    * Because `out` parameters break interface variance ([for the worst reasons...](https://stackoverflow.com/a/2877515/159145)) you can still support covariance of `TKey` (or is it contravariance?) by making it an extension-method on a variant interface, and expose `hasParent` through an explicit interface implementation.
2. Alternatively, define your own `struct Optional<T>`  which allows both `struct` and `class` types of `T` (whereas `System.Nullable<T>` only allows `struct` kinds of `T` that also cannot be `System.Nullable<T>`).

**Comment by ghost** [comment]
#leave-thread

**Comment by pikausp** [comment]
Is there any progress with this? Absolute killer feature :(

**Reply by jnm2** [reply]
C# 9 (released last November) has a feature for this. You can now write `T?` without constraining `T` to be a class or a struct. (In these cases, `T?` means "defaultable" rather than "nullable" since there's no metadata the compiler can emit for the latter which the runtime would understand.)

**Reply by daiplusplus** [reply]
@jnm2 That's disappointing - I was under the impression that `T?` would compile to `Nullable<T>` when `T : struct` or `T` when `T : class`. I tried it just now in LinqPad 6's beta and I'm disappointed now to see that with `class Foo<T> { public T? Value { get; set; } }` and `Foo<Int32> f = new Foo<Int32>();` that `( f.Value == ((Int32?)null) ) == false`.

This could be worked-around quite easily if .NET came with a simple `Option<T>` struct without any of the special rules and restrictions that `Nullable<T>` has (an `Option<T>` would be handy for JSON interop too, as it could represent `undefined` vs `null`).

**Reply by FaustVX** [reply]
For JSON, technically `undefined` is not a valid token.
[json.org](https://www.json.org/): under `value` group, only `null` is valid.

**Reply by huoyaoyuan** [reply]
> This could be worked-around quite easily if .NET came with a simple Option<T> struct

It was specifically designed not to be that. It's inefficient. Nullable reference type has no runtime overhead (or any runtime effect).

**Reply by daiplusplus** [reply]
@FaustVX Correct. By `undefined` I’m referring to when a JSON object lacks a property but is being deserialized to a class that has such a property. Currently JSON.Net and System.Text.Json both (by default) deserialise both cases to `null`. This makes things like `PATCH` DTOs harder to write. It’s not much work to roll your own JsonConverter to handle this, but it’s still extra work we shouldn’t have to be doing ourselves, imo.

**Reply by HaloFour** [reply]
I know that JSON.NET can be configured to recognize the difference and to support types like `Option<T>`.

**Reply by daiplusplus** [reply]
@huoyaoyuan I was not suggesting that `Option<T>` replace nullable reference types - I’m suggesting it as a way to easily and correctly handle nullable-reference-types *and* nullable-value-types in the same generic type.

That said, generic wrappers over value-types that are value-types themselves incur no significant overhead either, while not strictly zero-overhead the JIT reduces them down to a single blit copy, for which the cost is comparable to a vtable call or delegate invocation, which is negligible on computers made in the past 20+ years.

**Reply by daiplusplus** [reply]
@HaloFour Yes, it can - but you need to bring-your-own option type every time. There isn’t a standard one in `System.*` that we can all use and not have to reimplement in every new project.

**Reply by HaloFour** [reply]
I'm glad their isn't.  Any `Option<T>` type added to the BCL before the language and runtime explicitly work through the details of how DUs should work would almost invariably end up being wrong, kind of like how `System.Tuple<...>` didn't fit the bill when C# added tuples.  There are common [NuGet packages](https://www.nuget.org/packages/Optional/), though.

**Reply by daiplusplus** [reply]
`Option<T>` isn’t a DU though. While it can be thought of as a union of `T` with `void` or some other bottom-type, but C#/.NET does not support the concept of bottom-types, and none of the current DU proposals look like they’ll extend the type-system that way, so comparing it to a DU isn’t that useful in this particular case.

“Don’t let perfect be the enemy of good” - I know we don’t want to commit ourselves to supporting half-baked ideas for decades (**cough** `IList<T>`) but if there was something like Boost-for-.NET or an officially sanctioned `System2.*` library containing useful-but-imperfect types so we don’t need to go through dozens of third-party NuGet packages of unknown pedigree just to get something _that works_ out-the-door without reinventing the wheel every time… stuff like a Date struct, DateTime-with-ZoneId, IndentedTextWriter, Int128, and so on…

**Reply by HaloFour** [reply]
I'm not aware of a language that has a `Option<T>`/`Maybe<T>` type that isn't a union type.  Supporting such a type in C# with discriminated unions is certainly one of the considerations and it's mentioned several times in that proposal.

The idea of "sanctioned" packages is interesting but it feels like something that should be handled by the community and not Microsoft.  I wouldn't want any single entity to "enshrine" which third-party packages are preferred.

**Reply by jnm2** [reply]
@Jehoel Yes, everyone is disappointed. There's no sequence of bytes the C# compiler can emit to create an assembly with a signature that behaves like that though.

**Reply by kaleidocore** [reply]
> @Jehoel Yes, everyone is disappointed. There's no sequence of bytes the C# compiler can emit to create an assembly with a signature that behaves like that though.

This scenario is painfully common:
```
class Example<T> 
{
	public T? Value { get; set; }
}

void Main()
{
	var a = new Example<int>();
	a.Value = null; // Doesn't compile, but this is clearly what one would expect from the syntax above
}
```

This is a terrible shortcoming of C#. What about compiler workarounds? For example, could the compiler emit both value and reference implementations simultaneously under different (hidden) identifiers and later chose one implementation at concretion?

**Reply by CyrusNajmabadi** [reply]
this can lead to combinatorial explosion of generated symbols, which has it's own set of problems.
